storage: `StorageCollection` support + shard ID mappings

sploiselle opened the pull request:
### Motivation

This PR adds a known-desirable feature.
- Implements https://github.com/MaterializeInc/materialize/issues/14266#issuecomment-1255066383.
- Closes #14266

### Tips for reviewer

- @mjibson Pinging you just for the movement of `CoordTimestamp`
- This structure aims to make it easier to re-enable status shards, as well

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - there are no user-facing behavior changes

mjibson commented on src/repr/src/lib.rs:
```diff
@@ -78,6 +78,52 @@ pub struct Timestamp {
+
+impl TimestampManipulation for Timestamp {
+    fn step_forward(&self) -> Self {
+        match self.checked_add(1) {
+            Some(ts) => ts,
+            None => panic!("could not step forward"),
+        }
```
```suggestion
        self.step_forward()
```

mjibson commented on src/repr/src/lib.rs:
```diff
@@ -78,6 +78,52 @@ pub struct Timestamp {
+    }
+
+    fn step_forward_by(&self, amount: &Self) -> Self {
+        match self.checked_add(*amount) {
+            Some(ts) => ts,
+            None => panic!("could not step {self} forward by {amount}"),
+        }
```
```suggestion
        self.step_forward_by(amount)
```

mjibson commented on src/repr/src/lib.rs:
```diff
@@ -78,6 +78,52 @@ pub struct Timestamp {
+    }
+
+    fn step_back(&self) -> Option<Self> {
+        self.checked_sub(1)
```
```suggestion
        self.step_back()
```

mjibson reviewed (approved):
CoordTimestamp moving into a more common trait is exactly what I was hoping to get to during our discussion a few weeks ago. Thanks.

aljoscha reviewed (changes requested):
I like it!

I do have some recommendations for cleaning up the boundaries and stashing the new code into a more tightly abstracted thing:
 - I did a quick change that adds an `append_at_upper` to `persist_write_handles::PersistWorker`. That way, we can get around having to share the write upper using an `Arc`. See here: https://github.com/aljoscha/materialize/tree/pr-15052-shard-mapping
 -  I think we can stash the new managed collection worker behind an abstraction that should be quite similar to `persist_write_handles::PersistWorker`, that would move most of the code for the new task out of the storage controller itself and make it more self contained
 - we don't need to put the `persist_write_handles::PersistWorker` behind an `Arc` for sharing. It's a thin wrapper around an `UnboundedSender`, which should be cloneable. We can just make this cloneable, and give the new "managed task worker" that clone, so it can also just send commands to that queue
 - overall it would be pretty much the same code, but with tighter abstractions, I hope! 😅 

Also:
 - we should put the new collection into the `mz_internal` namespace and make sure that users cannot build views based on them (cc @teskje for how one goes about actually doing that 😅 )
 - and you didn't yet put in place the code that would re-initialize the whole thing when it notices a write conflict. I seem to remember that you mentioned this to me, right?

sploiselle commented on test/restart/mzcompose.py:
```diff
@@ -133,6 +135,38 @@ def workflow_stash(c: Composition) -> None:
+
+    # Storage collections are eventually consistent, so wait to be sure updates
+    # have made it.
+    time.sleep(2)
```
Oh sure! In my mind `testdrive` runs forever so adding 4 seconds wasn't material, but you're totally right that this should be better.

aljoscha commented:
> @aljoscha Your proposal for `AppendAtUpper` looks a lot like mine for `InfallibleAppend`: [sploiselle@4d3418a](https://github.com/sploiselle/materialize/commit/4d3418a4afb23a9489c3e5d13ec2b626d6236084)

ha! that looks excellent! It seems the only difference is that my change still periodically advances the upper to he latest timestamp we get from `now` but your's doesn't? Is that it?

Wouldn't we still need something that periodically advances the upper, to keep the table queryable at a recent timestamp? I might totally be missing sth...

> Forgot to mention this! But I don't think the above is necessary since you added the logic to PersistWorker to automatically retry indeterminate errors and proceed appropriately, i.e. when using PersistWorker you should never receive an indeterminate error. And the code holds enough other invariants about the collection's state that it never needs to worry about not having a sufficiently advanced understanding of the write frontier (it's the only writer).

That might be true! 😅

sploiselle commented:
@aljoscha `InfallibleAppend` puts a `NowFn` on the `PersistWorker` which drives the timestamp forward to essentially `max(now, current_write_frontier).step_forward()` on every update; the task then sends empty updates for all of its `GlobalId`s every tick. So that usecase is well-covered still 👍

benesch commented:
> > we should put the new collection into the mz_internal namespace and make sure that users cannot build views based on them (cc @teskje for how one goes about actually doing that 😅 )
> 
> You do this by defining their catalog items as not stable here:
> 
> https://github.com/MaterializeInc/materialize/blob/11d440e99f74b9759f7413da3dd3b4313efd96bd/src/adapter/src/catalog.rs#L261-L276

Heads up that I just put up a PR that proposes to tweak this logic a bit: https://github.com/MaterializeInc/materialize/pull/15089.

aljoscha commented:
> the task then sends empty updates for all of its `GlobalId`s every tick. So that usecase is well-covered still 👍

nice! 🙌 I didn't see that part. 

I didn't fully grasp those intricacies of the `infallible_append` approach, but I think it's good. It shouldn't be too hard to change that to `append_at_upper`, should we want to use a more "synchronized" clock in the future and/or want to provide serializability also across these introspection tables. Or we might never change it!

The one bikeshed I have is the name `infallible_append`. It can fail in case someone else writes to the shard. And it might fail for the "status shard" case where a `storaged` might be advancing the upper. Could maybe use sth like `append_at_now`, but I'm not super caught up on the name. 😅

sploiselle commented on src/ore/src/now.rs:
```diff
@@ -86,6 +86,30 @@ pub static SYSTEM_TIME: Lazy<NowFn> = Lazy::new(|| NowFn::from(system_time));
+    /// Return a monotonic `EpochMillis` value.
+    ///
+    /// Note that the minimum value this function ever returns is `1`.
+    pub fn now(&mut self) -> EpochMillis {
```
I removed `MonotonicNow` in favor of having the `PersistWriteHandle` offer a monotonic append that chooses all of the timestamps. Has a few advantages:
- We don't need to add retry logic for our writes bc currently we know we're the only writer to these collections.
- We don't need to track the maximum upper when adding new collections to avoid adding retry logic.

sploiselle commented:
@teskje ty for catching that; wound up moving @benesch 's PR into this one, and I hope that I've done so correctly. Can one of y'all check that [this](https://github.com/MaterializeInc/materialize/pull/15052/files#diff-fc176abe368c7d4fdcaf6efe5b25564b66994647baa39ca92d64df0aa8a23bb9L24) is correct? This change does not rely on replica data.

@aljoscha PTAL at commits starting with [storage: add CollectionManager trait](https://github.com/MaterializeInc/materialize/pull/15052/commits/dfff8d171064668703cbc132a77f7cec34db8e45).

I changed `append_at_upper`/`infallible_append` to `monotonic_append`, which takes an argument that says "make the monotonically increased timestamp at least this value." I think this design has a slightly clearer division of labor/is more appropriately configurable.

Also implemented abstractions you requested; much cleaner. 🙇

benesch commented:
@sploiselle heads up that I pushed up an [additional commit](https://github.com/MaterializeInc/materialize/pull/15052/commits/a641d20ce3a496ef486ad4a0b17972bf4be8d0e6) so that we can close out https://github.com/MaterializeInc/cloud/issues/3726:

> **adapter: add view to compute storage usage by object**
> The new mz_storage_usage view joins the underlying storage utilization
> data in mz_storage_usage_by_shard, which is collected by shard, with the
> new storage collection that describes the object_id <=> shard_id
> mapping.
> 
> This allows us to have a stable interface to storage usage information
> by object ID, even if we change how object IDs and shard IDs correspond
> in the future.

----

Separately, I noticed something funny. If you turn on `bin/environmentd` and try to query `mz_internal.mz_storage_shards`, it'll hang, seemingly forever. After a minute or two it seems to become queryable. Not sure what's up with that.

sploiselle commented:
@benesch Odd; I hadn't noticed that hanging. It does take a bit to become queryable because it has to ingest all of the persist data for the shards. If you had a long-running cluster, it's slow. I'll take a peek, though.

benesch commented:
Huh, you're right that it gets _way_ faster after blowing away whatever was in my `mzdata` directory.

> It does take a bit to become queryable because it has to ingest all of the persist data for the shards.

Wait, sorry, what is "it" here?

benesch commented on test/testdrive/catalog.td:
```diff
@@ -535,6 +532,7 @@ mz_scheduling_histogram_internal                log
 mz_scheduling_histogram_internal_1              log
 mz_scheduling_parks_internal                    log
 mz_scheduling_parks_internal_1                  log
+mz_storage_shards                               storage collection
```
I know this predates you, but we just _have_ to find a better name for this than "storage collection." In user facing stuff we should call it a "table", I think, since all builtin tables will look the same to the user. Internally I propose we call it a "StorageManagedTable" or something like that.

benesch replied:
Oh I don't care so much about the internal name—just want to avoid it leaking into the output of `SHOW SOURCES`!

aljoscha replied:
so we just want to make sure this shows up as `table` here, right?

sploiselle replied:
From conversation in the storage all hands today, the consensus is that these should be `source`s; I agree with that take so am going to run with it.

sploiselle commented:
@benesch re: https://github.com/MaterializeInc/materialize/pull/15052#issuecomment-1265433998 "it" is the shard mapping collection

teskje commented on test/sqllogictest/cluster_log_drop.slt:
```diff
@@ -21,7 +21,7 @@ mode cockroach
 SELECT (SELECT COUNT(*) FROM mz_catalog.mz_sources WHERE name NOT LIKE '%_1') - (SELECT COUNT(*) FROM mz_catalog.mz_sources WHERE name LIKE '%_1');
 ----
-0
+1
```
This LGTM. One alternative thing you could do to make this less annoying when adding more storage collections in the future is filter on `type = 'log'`.

aljoscha reviewed (approved):
This looks very good! I left some nits and comments and it would be good to resolve Nikhil's and Jan's comments as well.

Also, I'm assuming Matt's approving of the timestamp changes still holds, so didn't review that one too thoroughly.

Another also, we should somewhat quickly follow this up with renaming "storage collection" to sth more sensible, even if it's used only internally. And we might be able to add a test for the combined `mz_storage_usage` view.
